### PR TITLE
wip term

### PR DESCRIPTION
**Pull request by v8tenko**

### CODE REVIEW

File: __tests__/rules/diplodoc/__fixtures__/term.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+Определение термина может включать в себя базовую разметку.
+Листы:
+1. привет
+2. мир
+    1. я вложенный лист
+    2. и я
+- простой список
+    - вложенный
+
+
```
Comment (moki) [inline __tests__/rules/diplodoc/__fixtures__/term.ts:29]: two extra new lines where are they coming from?

File: __tests__/rules/diplodoc/__fixtures__/term.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        number: 1,
+    },
+    {
+        markdown: `\
+привет, мир 2
+
+[*ключ_термина]: Разъяснение термина или сокращения с помощью синтаксиса определений.
+Определение термина может включать в себя базовую разметку.
+Листы:
+1. привет
```
Comment (moki) [inline __tests__/rules/diplodoc/__fixtures__/term.ts:22]: you use lists in tests but you do not test what happens when term content is aligned with them(you should)

File: src/renderer/index.ts [hunk 3]
```diff
@@ -52,19 +53,28 @@ export type MarkdownRendererEnv = {
...
     ...list.initState(),
     ...GFMTable.initState(),
+    ...term.initState(),
 });
 
-class MarkdownRenderer<T = {}, CT extends ContainerBase = ContainerBase> extends CustomRenderer<T> {
+class MarkdownRenderer<
+    T = MarkdownRendererState,
+    CT extends ContainerBase = ContainerBase,
+> extends CustomRenderer<T> {
```
Comment (moki) [inline src/renderer/index.ts:77]: this is outside of the scope of this PR

File: src/rules/diplodoc/term.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    };
+};
+
+const initState = (): TermState => ({
+    term: {
+        pending: [],
+    },
+});
+
+const allways =
```
Comment (moki) [inline src/rules/diplodoc/term.ts:17]: its one l in always

File: src/rules/diplodoc/term.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        let rendered = ']';
+        const key = this.state.term.pending.pop();
+
+        rendered += `(*${key})`;
+
+        return rendered;
+    },
+    template_open: function (this: MarkdownRenderer<TermState>, tokens: Token[], i: number) {
+        const key = tokens[i].attrGet('label');
+        const rendered = this.EOL.repeat(2) + `[*${key}]: `;
```
Comment (moki) [inline src/rules/diplodoc/term.ts:44]: what happens when template is first in the token stream?

what happens when it is aligned with something other than paragraph?

you should have test cases for that and handle that appropriately.

File: src/rules/diplodoc/term.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+        return rendered;
+    },
+    template_close: function (this: MarkdownRenderer<TermState>) {
+        return this.EOL.repeat(2);
```
Comment (moki) [inline src/rules/diplodoc/term.ts:49]: same as above naive new line separation with everything wouldn't cut it

**Review by moki — CHANGES_REQUESTED** [review]

File: __tests__/rules/diplodoc/__fixtures__/term.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
+const term = [
+    {
+        markdown: `\
+[*ключ_термина]: Разъяснение термина или сокращения с помощью синтаксиса определений.
+Определение термина может включать в себя ##базовую## ~~разметку~~.
+
```
Comment (moki) [inline __tests__/rules/diplodoc/__fixtures__/term.ts:6]: this is an extra new line

check if "[*ключ_термина]:" token is last in the token stream, if so prevent rendering new line.

**Review by moki — CHANGES_REQUESTED** [review]
should term content "[*ключ_термина]: content..." be able to render inside the list item?

like this:

```
1. content
   [*ключ_термина]: content...
2. [*ключ_термина_1]: content...
```

theoretically since it is block container it should

we need test cases for it.

File: __tests__/rules/diplodoc/__fixtures__/term.ts (hunk 7 above)
Comment (v8tenko) [reply __tests__/rules/diplodoc/__fixtures__/term.ts:6]: first test is about it

Comment (moki) [reply __tests__/rules/diplodoc/__fixtures__/term.ts:6]: you have an extra new line in the end.

if this test colored green, then it means you have an error.

you should not have empty new line in the end.

File: src/rules/diplodoc/term.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        const key = tokens[i].attrGet('label');
+        const rendered = this.EOL.repeat(i ? 2 : 1) + `[*${key}]: `;
+
+        return rendered;
+    },
+    template_close: function (this: MarkdownRenderer<TermState>, tokens: Token[], i: number) {
+        const next = tokens[i + 1];
+        const isLast = i === tokens.length - 1;
+
+        if (!isLast && (!this.state.term.didRenderEOL || shouldSeparate(next))) {
```
Comment (moki) [inline src/rules/diplodoc/term.ts:55]: you can completely move logic if it should be separated with new line inside `shouldSeparate` function.

just `call`/`apply` with appropriate this context

File: src/rules/diplodoc/term.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    },
+    dfn_open: always(''),
+};
+
+function shouldSeparate(token: Token | undefined): boolean {
+    if (!token) {
+        return true;
+    }
+
+    if (token.type === 'bullet_list_close' || token.type === 'ordered_list_close') {
```
Comment (moki) [inline src/rules/diplodoc/term.ts:80]: these conditions could be merged.

**Pull request merged**